### ISSUE TITLE
<p><h4>hello</h4></p> misreported

### ISSUE THREAD

**Issue by alanfluff** [validator/validator#822]
Using this URL https://validator.w3.org/nu/#textarea

This code

```
<!DOCTYPE html>
<html lang="">
<head>
<title>Test</title>
</head>
<body>
<p><h4>hello</h4></p>
</body>
</html>
```

is reported with

```
Error: No p element in scope but a p end tag seen.
From line 8, column 18; to line 8, column 21
hello</h4></p>↩↩</bo
```

Remove the `h4` and all is green/good.

It surely should report about bad nesting (`h4` inside `p`)?

**Comment by Comandeer** [comment]
According to [tag omission rules for the `p` element](https://html.spec.whatwg.org/multipage/grouping-content.html#the-p-element), `p` element is automatically closed when several other HTML elements, including headings, are encountered. Therefore

```html
<p>
	<h4>Lorem ipsum</h4>
</p>
```

is in fact interpreted by a browser as

```html
<p></p>
<h4>Lorem ipsum</h4>
<p></p>
```

(which can be confirmed in devtools).

The error message is correct, as long as we consider HTML parsing rules: paragraph is closed before the heading and we're left with unpaired `</p>`. [The error message you mention is correct in such context in XML mode](https://validator.w3.org/nu/?doc=https%3A%2F%2Fbzdety.comandeer.pl%2Fparagraph.xhtml), as in XML mode we can actually insert heading into paragraph.

So it's not validator's fault, it's HTML weird parsing rules' fault ;)

**Comment by alanfluff** [comment]
Way cool explanation, thanks @Comandeer ^_^
